neptune/alpha/internal/utils/json_file_splitter.py

```diff
@@ -14,50 +14,75 @@
 # limitations under the License.
 #
 import json
+from collections import deque
 from io import StringIO
 from json import JSONDecodeError
 from typing import Optional
 
 
 class JsonFileSplitter:
 
-    # TODO: experiment with larger buffer sizes
     BUFFER_SIZE = 64 * 1024
+    MAX_PART_READ = 8 * 1024
 
     def __init__(self, file_path: str):
         self._file = open(file_path, "r")
         self._decoder = json.JSONDecoder(strict=False)
-        self._buffer = StringIO()
+        self._part_buffer = StringIO()
+        self._part_read = 0
+        self._parsed_queue = deque()
         self._start_pos = 0
 
     def close(self) -> None:
         self._file.close()
+        self._part_buffer.close()
 
     def get(self) -> Optional[dict]:
-        try:
-            return self._decode()
-        except (JSONDecodeError, ValueError):
-            self._read_data()
-            try:
-                return self._decode()
-            except (JSONDecodeError, ValueError):
-                return None
+        if self._parsed_queue:
+            return self._parsed_queue.popleft()
+        self._read_data()
+        if self._parsed_queue:
+            return self._parsed_queue.popleft()
 
     def _read_data(self):
-        if self._start_pos > 0:
-            new_buffer = StringIO()
-            new_buffer.write(self._buffer.getvalue()[self._start_pos:])
-            self._buffer.close()
-            self._buffer = new_buffer
-            self._start_pos = 0
-
-        data = self._file.read(self.BUFFER_SIZE)
-        while data:
-            self._buffer.write(data)
+        if self._part_read < self.MAX_PART_READ:
+            data = self._file.read(self.BUFFER_SIZE)
+            if not data:
+                return
+            if self._part_read > 0:
+                data = self._reset_part_buffer() + data
+            self._decode(data)
```

Comment:
There is a bug here I think. You release the date from the buffer, concatenate with new loaded data and store it in data variable. Then you pass it to the _decode method, which I think it is supposed to store the rest of unparsed data in the buffer since in 57-th line you overwrite data variable and previous data would be lost. However as I analyze _decode method, it does not always store rest of unparsed data, in case when first _json_start returns len(data).